projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v3-endpoint-blogging-prompts.php

```diff
@@ -0,0 +1,435 @@
+		add_action( 'pre_get_posts', array( $this, 'modify_query' ) );
+		add_filter( 'posts_clauses', array( $this, 'filter_sql' ) );
+		$items = parent::get_items( $request );
+		remove_filter( 'posts_clauses', array( $this, 'filter_sql' ) );
+		remove_action( 'pre_get_posts', array( $this, 'modify_query' ) );
+		restore_current_blog();
+
+		return $items;
+	}
+
+	/**
+	 * Retrieves a single blogging prompt.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		if ( ! $this->is_wpcom ) {
+			return $this->proxy_request_to_wpcom( $request, $request->get_param( 'id' ) );
+		}
+
+		switch_to_blog( self::TEMPLATE_BLOG_ID );
+		$item = parent::get_item( $request );
+		restore_current_blog();
+
+		return $item;
+	}
+
+	/**
+	 * Modify the posts query using the {@see 'pre_get_posts'} hook.
+	 *
+	 * @param WP_Query $wp_query The WP_Query instance (passed by reference).
+	 */
+	public function modify_query( &$wp_query ) {
+		if ( is_array( $wp_query->query_vars['date_query'] ) ) {
+			$wp_query->query_vars['date_query'] = array_map(
+				array( $this, 'map_date_query' ),
+				$wp_query->query_vars['date_query']
+			);
+		}
+	}
+
+	/**
+	 * Modify date_query items when querying prompts.
+	 *
+	 * @link https://developer.wordpress.org/reference/classes/WP_Query/#date-parameters
+	 *
+	 * @param  array|string|null $date_query Date query.
+	 * @return array|string|null             Modified date query.
+	 */
+	public function map_date_query( $date_query ) {
+		if ( isset( $date_query['after'] ) ) {
+			// `after` date queries should include posts on the specified date, so force `inclusive` queries.
+			$date_query['inclusive'] = true;
+
+			// If using a "year-less" date, e.g. `--03-16`, override the date_query, and prepare to modify sql manually.
+			// `after` should be a date string when making API requests, rather than an array.
+			if ( is_string( $date_query['after'] ) && 0 === strpos( $date_query['after'], '-' ) ) {
+				$date = date_create_from_format( '--m-d', $date_query['after'] );
+
+				if ( false !== $date ) {
+					// PHP day of the year starts with 0; normalize to match SQL DAYOFTHEYEAR which starts with 1.
+					$this->day_of_year_query = $date->format( 'z' ) + 1;
+
+					// Unset the date query, since we'll by modifying the SQL manually.
+					return null;
+				}
+			}
+		}
+
+		return $date_query;
+	}
+
+	/**
+	 * Modify post sql for custom date ordering using the {@see 'posts_clauses'} hook.
+	 *
+	 * @param array $clauses SQL clauses for the current query.
+	 * @return array         Modified SQL clauses.
+	 */
+	public function filter_sql( $clauses ) {
+		global $wpdb;
+		if ( $this->day_of_year_query > 0 ) {
+			$day = $this->day_of_year_query;
+
+			// Grab the current sort order, `ASC` or `DESC`, so we can reuse it.
+			$order = end( explode( ' ', $clauses['orderby'] ) );
```

Comment:
When I pass the `order=desc` param from the posts controller, I expected this to reverse the order of the posts, but they stayed the same. I notice that the `order` rest param may be `desc` or `asc` but the `orderby` param is the field name such as `date` or `id`. Is that the issue? https://developer.wordpress.org/rest-api/reference/posts/#arguments

Maybe I' not understanding the intent of this code, I don't think we do even need to support `order` or `orderby` anyway?

Reply:
Thanks for taking a look @roo2 !

The `order` and `orderby` param work as normal by themselves. When you add the yearless `after` param (e.g. `--02-28`), the order param affects the order of posts _within each day_. We don't see this with the current prompts, since there are only one per day, but if you had multiple prompts for the same day, but different years, this would control whether the most recent (`desc`) or oldest (`asc`) will come first for that day.

The tests for this in the original wpcom diff (D102655-code) help illustrate, take a look at
- `test_blogging_prompts_collection_orders_by_month_and_day`
- `test_blogging_prompts_collection_orders_by_month_and_day_desc`

Using the `order` param this way seemed much more useful than reverse ordering the days, themselves. I'm assuming if you provided an after param with a day, you probably want those posts at the top, since that's the date you care about. But you may want to get the most recent prompt for that day (`order=desc`, like what I'm doing in #29189), or you might want a list with oldest to newest to cycle through all the prompts for that day. This way provides that flexibility.